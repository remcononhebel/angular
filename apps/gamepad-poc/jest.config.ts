/* eslint-disable */
export default {
  displayName: 'gamepad-poc',
  preset: '../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  globals: {},
  coverageDirectory: '../../coverage/apps/gamepad-poc',
  coveragePathIgnorePatterns: ['.jest-mock.ts'],
  transform: {
    '^.+\\.(ts|mjs|js|html)$': [
      'jest-preset-angular',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\.(html|svg)$',
      },
    ],
  },
  transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
};
