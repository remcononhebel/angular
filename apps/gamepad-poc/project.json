{
  "name": "gamepad-poc",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "application",
  "prefix": "app",
  "sourceRoot": "apps/gamepad-poc/src",
  "tags": [],
  "targets": {
    "build": {
      "executor": "@angular-devkit/build-angular:application",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/apps/gamepad-poc",
        "index": "apps/gamepad-poc/src/index.html",
        "browser": "apps/gamepad-poc/src/main.ts",
        "polyfills": ["zone.js"],
        "tsConfig": "apps/gamepad-poc/tsconfig.app.json",
        "inlineStyleLanguage": "scss",
        "assets": [
          "apps/gamepad-poc/src/favicon.ico",
          "apps/gamepad-poc/src/assets"
        ],
        "styles": ["apps/gamepad-poc/src/styles.scss"],
        "scripts": []
      },
      "configurations": {
        "production": {
          "budgets": [
            {
              "type": "initial",
              "maximumWarning": "500kb",
              "maximumError": "1mb"
            },
            {
              "type": "anyComponentStyle",
              "maximumWarning": "2kb",
              "maximumError": "4kb"
            }
          ],
          "outputHashing": "all"
        },
        "development": {
          "optimization": false,
          "extractLicenses": false,
          "sourceMap": true
        }
      },
      "defaultConfiguration": "production"
    },
    "serve": {
      "executor": "@angular-devkit/build-angular:dev-server",
      "configurations": {
        "production": {
          "buildTarget": "gamepad-poc:build:production"
        },
        "development": {
          "buildTarget": "gamepad-poc:build:development"
        }
      },
      "defaultConfiguration": "development"
    },
    "extract-i18n": {
      "executor": "@angular-devkit/build-angular:extract-i18n",
      "options": {
        "buildTarget": "gamepad-poc:build"
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"]
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "apps/gamepad-poc/jest.config.ts"
      }
    },
    "serve-static": {
      "executor": "@nx/web:file-server",
      "options": {
        "buildTarget": "gamepad-poc:build",
        "staticFilePath": "dist/apps/gamepad-poc/browser"
      }
    }
  }
}
