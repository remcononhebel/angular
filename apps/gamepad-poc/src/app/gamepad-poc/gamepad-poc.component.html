<div class="debug">
  <div class="debug__joystick-info">
    <h2>Left joystick</h2>
    <pre>{{ left$ | async | json }}</pre>
  </div>

  <div class="debug__joystick-info">
    <h2>Right joystick</h2>
    <pre>{{ right$ | async | json }}</pre>
  </div>
</div>

<div class="joysticks">
  <app-joystick
    [joystickId]="'leftJoystick'"
    [options]="leftJoystickOptions"
    (move)="onLeftJoystickMove($event)"></app-joystick>

  <app-joystick
    [joystickId]="'rightJoystick'"
    [options]="rightJoystickManagerOptions"
    (move)="onRightJoystickMove($event)"></app-joystick>
</div>
