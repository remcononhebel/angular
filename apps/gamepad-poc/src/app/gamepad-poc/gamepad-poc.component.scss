.debug {
  display: flex;
  justify-content: space-between;

  &__joystick-info {
    background-color: #ddd;
    border: 1px solid black;
    padding: 8px;
    width: 50%;
  }
}

.joysticks {
  position: absolute;
  bottom: 0;
  width: 100%;
  display: flex;
  justify-content: space-between;
}
