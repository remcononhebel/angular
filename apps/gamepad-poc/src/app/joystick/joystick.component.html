<div class="joystick-container">
  <div [attr.id]="joystickId">&nbsp;</div>
</div>
