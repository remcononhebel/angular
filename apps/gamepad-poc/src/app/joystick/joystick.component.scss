.joystick-container {
  border: 1px solid black;
  height: 200px;
  position: relative;
  z-index: 1000;
  width: 200px;
}

.joystick {
  height: 100%;
}

.joystick {
}
