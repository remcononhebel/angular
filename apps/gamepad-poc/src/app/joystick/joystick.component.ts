import { AfterViewInit, Component, EventEmitter, Input, Output } from '@angular/core';
import nipplejs from 'nipplejs';

@Component({
  selector: 'app-joystick',
  standalone: true,
  imports: [],
  templateUrl: './joystick.component.html',
  styleUrl: './joystick.component.scss',
})
export class JoystickComponent implements AfterViewInit {
  @Input() joystickId!: string;
  @Input() options!: nipplejs.JoystickManagerOptions;

  @Output() move = new EventEmitter<nipplejs.JoystickOutputData>();

  private manager!: nipplejs.JoystickManager;

  ngAfterViewInit(): void {
    const options: nipplejs.JoystickManagerOptions = { ...this.options, zone: this.getJoystickElement(this.joystickId) };

    this.manager = nipplejs.create(options);
    this.manager.on('move', (_: unknown, data: nipplejs.JoystickOutputData) => {
      this.move.emit(data);
    });
  }

  getJoystickElement(joystickId: string): HTMLElement | undefined {
    return document.getElementById(joystickId) ?? undefined;
  }
}
