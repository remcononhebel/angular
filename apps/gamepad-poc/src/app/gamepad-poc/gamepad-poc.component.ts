import { CommonModule } from '@angular/common';
import { Component } from '@angular/core';
import nipplejs from 'nipplejs';
import { Subject, map } from 'rxjs';
import { JoystickComponent } from '../joystick/joystick.component';

export interface JoystickViewModel {
  direction: nipplejs.Direction;
}

@Component({
  selector: 'app-gamepad-poc',
  standalone: true,
  imports: [CommonModule, JoystickComponent],
  templateUrl: './gamepad-poc.component.html',
  styleUrl: './gamepad-poc.component.scss',
})
export class GamepadPocComponent {
  readonly leftJoystickOutputData$ = new Subject<nipplejs.JoystickOutputData>();
  readonly rightJoystickOutputData$ = new Subject<nipplejs.JoystickOutputData>();

  readonly left$ = this.leftJoystickOutputData$.pipe(map((data) => this.mapToViewModel(data)));
  readonly right$ = this.rightJoystickOutputData$.pipe(map((data) => this.mapToViewModel(data)));

  leftJoystickManager?: nipplejs.JoystickManager;
  rightJoystickManager?: nipplejs.JoystickManager;

  readonly leftJoystickOptions: nipplejs.JoystickManagerOptions = {
    mode: 'static',
    position: { left: '100px', bottom: '100px' },
    color: 'red',
  };

  readonly rightJoystickManagerOptions: nipplejs.JoystickManagerOptions = {
    mode: 'static',
    position: { left: '100px', bottom: '100px' },
    color: 'blue',
  };

  onLeftJoystickMove(event: nipplejs.JoystickOutputData) {
    this.leftJoystickOutputData$.next(event);
  }

  onRightJoystickMove(event: nipplejs.JoystickOutputData) {
    this.rightJoystickOutputData$.next(event);
  }

  private mapToViewModel(data: nipplejs.JoystickOutputData): JoystickViewModel | null {
    return { direction: data.direction };
  }
}
