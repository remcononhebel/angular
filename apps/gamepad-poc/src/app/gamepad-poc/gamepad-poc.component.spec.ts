import { ComponentFixture } from '@angular/core/testing';
import { Spectator, createComponentFactory } from '@ngneat/spectator';
import { MockComponent } from 'ng-mocks';
import nipplejs from 'nipplejs';
import { JoystickComponent } from '../joystick/joystick.component';
import { JoystickPocComponent } from './gamepad-poc.component';

const mockJoystickOutputData: nipplejs.JoystickOutputData = {
  direction: {
    angle: 'up',
    x: 'left',
    y: 'up',
  },
} as nipplejs.JoystickOutputData;

describe('JoystickPocComponent', () => {
  let spectator: Spectator<JoystickPocComponent>;
  let component: JoystickPocComponent;
  let fixture: ComponentFixture<JoystickPocComponent>;

  const createComponent = createComponentFactory({
    component: JoystickPocComponent,
    imports: [MockComponent(JoystickComponent)],
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
    fixture = spectator.fixture;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render two joysticks and two debug info divs', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('onLeftJoystickMove', () => {
    it('should emit leftJoystickOutputData$', (done: jest.DoneCallback) => {
      component.leftJoystickOutputData$.subscribe((data: nipplejs.JoystickOutputData) => {
        expect(data).toEqual(mockJoystickOutputData);
        done();
      });

      component.onLeftJoystickMove(mockJoystickOutputData);
    });
  });

  describe('onRightJoystickMove', () => {
    it('should emit rightJoystickOutputData$', (done: jest.DoneCallback) => {
      component.rightJoystickOutputData$.subscribe((data: nipplejs.JoystickOutputData) => {
        expect(data).toEqual(mockJoystickOutputData);
        done();
      });

      component.onRightJoystickMove(mockJoystickOutputData);
    });
  });

  it('should have the correct input value for the child component', () => {
    const child = spectator.query(JoystickComponent);
    expect(child).toBeTruthy();
    if (child) {
      child.joystickId = 'yes';
      spectator.detectChanges();
      expect(fixture).toMatchSnapshot();
    } else {
      fail();
    }
  });
});
