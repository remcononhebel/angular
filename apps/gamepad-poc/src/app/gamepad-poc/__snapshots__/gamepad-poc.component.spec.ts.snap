// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoystickPocComponent should have the correct input value for the child component 1`] = `
<app-gamepad-poc
  left$={[Function AnonymousSubject]}
  leftJoystickOptions={[Function Object]}
  leftJoystickOutputData$={[Function Subject]}
  right$={[Function AnonymousSubject]}
  rightJoystickManagerOptions={[Function Object]}
  rightJoystickOutputData$={[Function Subject]}
>
  <div
    class="debug"
  >
    <div
      class="debug__joystick-info"
    >
      <h2>
        Left joystick
      </h2>
      <pre>
        null
      </pre>
    </div>
    <div
      class="debug__joystick-info"
    >
      <h2>
        Right joystick
      </h2>
      <pre>
        null
      </pre>
    </div>
  </div><div
    class="joysticks"
  >
    <app-joystick />
    <app-joystick />
  </div>
</app-gamepad-poc>
`;

exports[`JoystickPocComponent should render two joysticks and two debug info divs 1`] = `
<app-gamepad-poc
  left$={[Function AnonymousSubject]}
  leftJoystickOptions={[Function Object]}
  leftJoystickOutputData$={[Function Subject]}
  right$={[Function AnonymousSubject]}
  rightJoystickManagerOptions={[Function Object]}
  rightJoystickOutputData$={[Function Subject]}
>
  <div
    class="debug"
  >
    <div
      class="debug__joystick-info"
    >
      <h2>
        Left joystick
      </h2>
      <pre>
        null
      </pre>
    </div>
    <div
      class="debug__joystick-info"
    >
      <h2>
        Right joystick
      </h2>
      <pre>
        null
      </pre>
    </div>
  </div><div
    class="joysticks"
  >
    <app-joystick />
    <app-joystick />
  </div>
</app-gamepad-poc>
`;
